Jenkinsfile

```diff
@@ -0,0 +1,614 @@
+    """ : "")
+  }
+
+  environment {
+    OS_VERSION = 'ubuntu/bionic64/java-driver'
+    JABBA_SHELL = '/home/jenkins/.jabba/jabba.sh'
+    CCM_ENVIRONMENT_SHELL = '/usr/local/bin/ccm_environment.sh'
+    SERIAL_ITS_ARGUMENT = "-DskipSerialITs=${params.SKIP_SERIAL_ITS}"
+    ISOLATED_ITS_ARGUMENT = "-DskipIsolatedITs=${params.SKIP_ISOLATED_ITS}"
+    PARALLELIZABLE_ITS_ARGUMENT = "-DskipParallelizableITs=${params.SKIP_PARALLELIZABLE_ITS}"
+    INTEGRATION_TESTS_FILTER = "${params.INTEGRATION_TESTS_FILTER}"
+  }
+
+  stages {
+    stage ('Per-Commit') {
+      options {
+        timeout(time: 2, unit: 'HOURS')
+      }
+      when {
+        beforeAgent true
+        allOf {
+          expression { params.ADHOC_BUILD_TYPE == 'BUILD' }
+          expression { params.CI_SCHEDULE == 'DO-NOT-CHANGE-THIS-SELECTION' }
+          expression { params.CI_SCHEDULE_SERVER_VERSIONS == 'DO-NOT-CHANGE-THIS-SELECTION' }
+          expression { params.CI_SCHEDULE_JABBA_VERSION == 'DO-NOT-CHANGE-THIS-SELECTION' }
+          not { buildingTag() }
+        }
+      }
+
+      matrix {
+        axes {
+          axis {
+            name 'SERVER_VERSION'
+            values '3.11',     // Latest stable Apache CassandraⓇ
+                   '4.0',      // Development Apache CassandraⓇ
+                   'dse-6.8.0' // Current DataStax Enterprise
+          }
+        }
+
+        agent {
+          label "${OS_VERSION}"
+        }
+        environment {
+          // Per-commit builds are only going to run against JDK8
+          JABBA_VERSION = '1.8'
+        }
+
+        stages {
+          stage('Initialize-Environment') {
+            steps {
+              initializeEnvironment()
+              script {
+                if (env.BUILD_STATED_SLACK_NOTIFIED != 'true') {
+                  notifySlack()
+                }
+              }
+            }
+          }
+          stage('Describe-Build') {
+            steps {
+              describePerCommitStage()
+            }
+          }
+          stage('Build-Driver') {
+            steps {
+              buildDriver(env.JABBA_VERSION)
+            }
+          }
+          stage('Execute-Tests') {
+            steps {
+              catchError {
+                // Use the matrix JDK for testing
+                executeTests()
+              }
+            }
+            post {
+              always {
+                /*
+                 * Empty results are possible
+                 *
+                 *  - Build failures during mvn verify may exist so report may not be available
+                 */
+                junit testResults: '**/target/surefire-reports/TEST-*.xml', allowEmptyResults: true
+                junit testResults: '**/target/failsafe-reports/TEST-*.xml', allowEmptyResults: true
+              }
+            }
+          }
+          stage('Execute-Code-Coverage') {
+            // Ensure the code coverage is run only once per-commit
```

Comment:
Code coverage, as far as I understand, is being performed for all runs, but we are only collecting and displaying results for SERVER_VERSION = 4.0. Not sure if it's useful to limit the coverage displaying without limiting the coverage performing.
Also, I'm not seeing code coverage results anywhere in Blue Ocean. Not sure if they are being correctly collected.

Reply:
Code coverage has a few issues with this setup. Since each combination is run on a separate VM, aggregating the results is going to be a challenge. Also, the JaCoCo plugin for CloudBees is on version 3.0.4, which does not collect the coverage metrics and make the charts in Jenkins if any phase of the pipeline reports as failed or unstable. So unless every test run passes, the coverage metrics aren't collected. Version 3.0.5 of the plugin allows for forcing the collection to happen, even if tests fail. That version of the plugin is not yet available via the CloudBees installation, I can look into upgrading it.

The metrics have been collected for a few of the runs that completed with all tests passing, but I think you can only see them on the regular Jenkins UI (i.e. not Blue Ocean). You can see the coverage trend here:
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/jacoco/
And the last run here:
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/40/
and
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/40/jacoco

TL;DR; The code coverage stuff needs more attention.

Reply:
> That version of the plugin is not yet available via the CloudBees installation, I can look into upgrading it.

Thanks for the explanations, would you mind filing a Jira for the code coverage issues and details what you think should/could be improved?

Also, I was looking at the dashboards you linked, the coverage numbers seem excessively low. We might want to check their correctness by comparing with a local build.

Reply:
Jira filed here: https://datastax-oss.atlassian.net/browse/JAVA-2724.
I'll check the coverage we are currently getting with a build/test against 4.0 locally and on CloudBees.

Reply:
Thank you!